framework/db/pgsql/ArrayExpression.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace yii\db\pgsql;
+
+use yii\db\ExpressionInterface;
+use yii\db\Query;
+use yii\db\QueryBuilder;
+use yii\db\QueryInterface;
+
+/**
+ * ArrayExpression represents a SQL expression that represents a PostreSQL array.
+ *
+ * Expressions of this type can be used for example in conditions, like:
+ *
+ * ```php
+ * $query->andWhere(['@>', 'items', new ArrayExpression([1, 2, 3], 'integer')])
+ * ```
+ *
+ * which will result in a condition `WHERE "items" @> ARRAY[1, 2, 3]::integer[]`.
+ *
+ * @see https://www.postgresql.org/docs/9.6/static/arrays.html
+ * @see https://www.postgresql.org/docs/current/static/functions-array.html
+ * @author Dmitry Naumenko <d.naumenko.a@gmail.com>
+ * @since 2.0.13
+ */
+class ArrayExpression implements ExpressionInterface
+{
+    const PARAM_PREFIX = ':qp';
+
+    /**
+     * @var null|string the type of the array elements. Defaults to `null` which means the type is
+     * not explicitly specified. This may result in an error if the type can not be inferred from the context.
+     * @see https://www.postgresql.org/docs/9.6/static/arrays.html
+     */
+    protected $type;
+    /**
+     * @var array|QueryInterface|mixed the array content. Either represented as an array of values or a Query that
+     * returns these values. A single value will be considered as an array containing one element.
+     */
+    protected $values;
+
+
+    /**
+     * ArrayExpression constructor.
+     *
+     * @param array|QueryInterface|mixed $values the array content. Either represented as an array of values or a Query that
+     * returns these values. A single value will be considered as an array containing one element.
+     * @param string|null $type the type of the array elements. Defaults to `null` which means the type is
+     * not explicitly specified. This may result in an error if the type can not be inferred from the context.
+     */
+    public function __construct($values, $type = null)
+    {
+        $this->values = $values;
+        $this->type = $type;
+    }
+
+    /**
+     * @return string the typecast expression based on [[type]].
+     */
+    protected function getTypecast()
+    {
+        if ($this->type === null) {
+            return '';
+        }
+
+        $result = '::' . $this->type;
+        if (strpos($this->type, '[]') === false) {
+            $result .= '[]';
+        }
+
+        return $result;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function buildUsing(QueryBuilder $queryBuilder, &$params = [])
+    {
+        $value = $this->values;
+
+        if ($value instanceof Query) {
+            list ($sql, $params) = $queryBuilder->build($value, $params);
```

Comment:
It is better to introduce check for `instanceof Query` somewhere inside `QueryBuilder`.
Having `Query` instance as one of the argument affects ay condtion composition, for example `IN` SQL operator.

Reply:
Than it is impossible to build condition like this:

```
column && array[(SELECT id FROM table), (SELECT id FROM another_table)]
```

Reply:
I would solve this using some kind of placeholder mechanism:

```php
interface ExpressionInterface
{
    public function getSubQueries();
}

class ArrayExpression implements ExpressionInterface
{
   public function getSql()
   {
       return column && array[({[q1]}), ({[q1]})];
   {

    public function getSubQueries()
    {
        return [
            '{[q1]}' => new Query()->setSql('SELECT id FROM table'),
            '{[q2]}' => new Query()->setSql('SELECT id FROM another_table'),
        ];
    }
}
```

Reply:
Or there should setter and getter for `yii\db\Connection` instance inside expression, so by default it will pick `Yii::$app->db`, get query builder from it and solve the issue.

Reply:
> Or there should setter and getter for yii\db\Connection instance inside expression, so by default it will pick Yii::$app->db, get query builder from it and solve the issue.

this does not make sense, an expression is always part of some sql query, which should use the connection and query builder in that context and not try to fallback to global db connection.

Reply:
The problem is this PR twist the original idea of `yii\db\Expression`. Expression was created to bypass parameters binding, allowing to paste SQL statements instead of string param at particular place.
This PR considers `Expression` is a light-weight version of `yii\db\Query`, which should be built by `QueryBuilder` inside its owner `Query` object.
I would say that `yii\db\Query` should extend `yii\db\Expression` creating a 'Composite' pattern, allowing building of nested queries with unlimited nesting level.

I would rather allow array condition to be specified via `yii\db\Query` instance (or its descendant) rather then `Expression`.

Reply:
Instead of introducing `build()` method inside `Expression` and gess a good name for it - it is better to adjsut `QueryBuilder::build()` in the way it can process both `Query` and `Expression` and use it as universal interface.

Reply:
> Instead of introducing build() method inside Expression and gess a good name for it - it is 
> better to adjsut QueryBuilder::build() in the way it can process both Query and Expression 
> and use it as universal interface.

I was thinking about this too but we don't have a Query or expression here, but a nesting, where a Query object resides inside of Expression. 

```php
$query->where(['@>', 'somecolumn', new ArrayExpression([(new Query())->...], 'integer')]);
```

So we need some way to add subqueries to the Expression so that Querybuilder can find it.
One way to allow this would be to allow Query() objects inside parameters, so that paramters that are a Query, would be replaced by the Query SQL string. But we need to check the consequences of supporting this first.

Reply:
We still have issues with subqueries not being useable everywhere, like
`['in', 'id', [(new Query())->select(1), (new Query())->select(2)]]`
or even worse
`['in', ['id', 'id2'], [['id' => (new Query())->select(1), 'id2' => (new Query())->select(2)]]]`

Implementing it everywhere will be pain, solution to this is not obvious.

If we stick to this approach we'll have more complex code in most methods of Query builder.

We could pass this PR without subquery features, make
```php
interface ExpressionInterface
{
    …
    /**
     * @internal currently not used, introduced to avoid backward-incompatible changes, may be dropped
     */
    public function getSubQueries();
}
```

On the other I like the proposal by @cebe to keep subqueries in $params, as we can easily handle those at the end. Then I would make `Query` implements `ExpressionInterface`, as it should make QueryBuilder simpler and robust.

Reply:
> The problem is this PR twist the original idea of yii\db\Expression.

Maybe `QueryPartInterface` would be more suitable then `ExpressionInterface` if we decide to put there features like subqueries.